src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,226 @@
+
+    // I
+    simplePid.integral += altErrorCm * taskIntervalSeconds; // cm * seconds
+    float iOut = simplePid.ki * simplePid.integral;
+    // arbitrary limit on iTerm, same as for gps_rescue, +/-20% throttle
+    iOut = constrainf(iOut, -200.0f, 200.0f); 
+
+    // D
+    const float derivative = (simplePid.lastAltitude - altHoldState.measuredAltitudeCm) / taskIntervalSeconds; // cm/s
+    simplePid.lastAltitude = altHoldState.measuredAltitudeCm;
+    float dOut = simplePid.kd * derivative;
+    // PT2 filtering at altitude_d_lpf / 100 cutoff, default 1s, the same as for GPS Rescue 
+    dOut = pt2FilterApply(&altHoldDeltaLpf, dOut);
+
+    // F
+    const float fOut = altholdConfig()->altHoldPidD * altHoldState.targetAltitudeDelta;
+    // if error is used, we get a 'free kick' in derivative from changes in the target value
+    // but this is delayed by the smoothing, leading to lag and overshoot.
+    // calculating feedforward separately avoids the filter lag.
+
+    const float output = pOut + iOut + dOut + fOut;
+
+    DEBUG_SET(DEBUG_ALTHOLD, 3, (lrintf)(output));
+    DEBUG_SET(DEBUG_ALTHOLD, 4, (lrintf)(pOut));
+    DEBUG_SET(DEBUG_ALTHOLD, 5, (lrintf)(iOut));
+    DEBUG_SET(DEBUG_ALTHOLD, 6, (lrintf)(dOut));
+    DEBUG_SET(DEBUG_ALTHOLD, 7, (lrintf)(fOut));
+
+    return output;
+}
+
+void simplePidInit(float kp, float ki, float kd)
+{
+    simplePid.kp = kp;
+    simplePid.ki = ki;
+    simplePid.kd = kd;
+    simplePid.lastAltitude = 0.0f;
+    simplePid.integral = 0.0f;
+}
+
+void altHoldReset(void)
+{
+    altHoldState.targetAltitudeCm = getAltitude(); // current altitude in cm
+    simplePid.lastAltitude = altHoldState.measuredAltitudeCm;
+    simplePid.integral = 0.0f;
+    altHoldState.throttleOut = positionConfig()->hover_throttle;
+    altHoldState.targetAltitudeDelta = 0.0f;
+}
+
+void altHoldInit(void)
+{
+    simplePidInit(
+        ALT_HOLD_PID_P_GAIN * altholdConfig()->altHoldPidP,
+        ALT_HOLD_PID_I_GAIN * altholdConfig()->altHoldPidI,
+        ALT_HOLD_PID_D_GAIN * altholdConfig()->altHoldPidD);
+        // the multipliers are base scale factors
+        // iTerm is relatively weak, intended to be slow moving to adjust baseline errors
+        // the Hover value is important otherwise takes time for iTerm to correct
+        // High P will wobble readily
+        // with these scalers, the same numbers as for GPS Rescue work OK for altHold
+        // the goal is to share these gain factors, if practical for all altitude controllers
+
+    //setup altitude D filter
+    const float cutoffHz = 0.01f * positionConfig()->altitude_d_lpf; // default 1Hz, time constant about 160ms
+    const float gain = pt2FilterGain(cutoffHz, taskIntervalSeconds);
+    pt2FilterInit(&altHoldDeltaLpf, gain); 
+
+    // similar PT2 on accelerometer to remove noise
+    const float gainAcc = pt2FilterGain((float)accelerometerConfig()->acc_lpf_hz, taskIntervalSeconds);
+    pt2FilterInit(&altHoldAccVerticalLpf, gainAcc);
```

Comment:
```suggestion
```
left from acc code?

Reply:
Thank you, I didn't notice that I forgot to remove them.